### ISSUE TITLE
Checkout: Invalid memory addres or nil pointer dereference

### ISSUE THREAD

**Issue by mister-turtle** [go-git/go-git#182]
**Versions:**
github.com/go-git/go-billy/v5 v5.0.0
github.com/go-git/go-git/v5 v5.2.0

**Issue**
This involves a repo in a bit of an odd state, as I'm writing a tool to retrieve erroneously exposed git repositories from web servers. Pulling down as much of the external repository as possible will produce local repository where a file is both staged for addition to the repository, is deleted as an unstaged commit, but and not present on the filesystem. After running the tool to retrieve the remote .git directory:

```shell
$ git status
On branch master
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        new file:   file1.md
        new file:   file2.md

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        deleted:    file1.md
        deleted:    file2.md
```

Using git cli, allows us to checkout a single file, resulting in the file being available on the filesystem.

```shell
$ ls -l
total 0

$ git checkout file1.md

$ ls -l
total 4
-rw-r--r-- 1 tester tester 7 Oct 12 09:59 file1.md

$ git checkout file2.md

$ ls -l
total 8
-rw-r--r-- 1 tester tester 7 Oct 12 09:59 file1.md
-rw-r--r-- 1 tester tester 6 Oct 12 10:01 file2.md
```

However, using go-git, we cannot checkout individual files, and attempting to checkout will produce the following results:

```go
fs := osfs.New(filepath.Join(g.OutputDir, ".git"))
s := filesystem.NewStorage(fs, nil)
r, err := git.Open(s, fs)
if err != nil {
	return err
}

err = tree.Checkout(&git.CheckoutOptions{})
if err != nil {
	return err
}
		

Output
Error: worktree contains unstaged changes
```

Providing the Force option to the git checkout options gives us our nil pointer:
```go
fs := osfs.New(filepath.Join(g.OutputDir, ".git"))
s := filesystem.NewStorage(fs, nil)
r, err := git.Open(s, fs)
if err != nil {
	return err
}

err = tree.Checkout(&git.CheckoutOptions{
	Force: true,
})


Output
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x718092]

goroutine 1 [running]:
github.com/go-git/go-git/v5/storage/filesystem.(*ObjectStorage).getFromUnpacked(0xc0001826f8, 0x94e8ebf2662c7d0e, 0xa61b59e510c38080, 0x13c0c7fe, 0x0, 0x0, 0x0, 0x0)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/storage/filesystem/object.go:388 +0x1c2
github.com/go-git/go-git/v5/storage/filesystem.(*ObjectStorage).EncodedObject(0xc0001826f8, 0xe8ebf2662c7d0e01, 0x1b59e510c3808094, 0x13c0c7fea6, 0x0, 0x0, 0x0, 0x0)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/storage/filesystem/object.go:322 +0x3e8
github.com/go-git/go-git/v5/plumbing/object.GetCommit(0x7f0a395ff328, 0xc0001826e0, 0x94e8ebf2662c7d0e, 0xa61b59e510c38080, 0xc013c0c7fe, 0x14, 0x662c7d0e00127000, 0x10c3808094e8ebf2)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/plumbing/object/commit.go:55 +0x4e
github.com/go-git/go-git/v5.(*Repository).CommitObject(...)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/repository.go:1334
github.com/go-git/go-git/v5.(*Worktree).getTreeFromCommitHash(0xc000139050, 0x94e8ebf2662c7d0e, 0xa61b59e510c38080, 0x13c0c7fe, 0x0, 0x0, 0x0)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/worktree.go:629 +0x91
github.com/go-git/go-git/v5.(*Worktree).Reset(0xc000139050, 0xc00011fab3, 0x11, 0x94e8ebf2662c7d0e)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/worktree.go:288 +0xf0
github.com/go-git/go-git/v5.(*Worktree).Checkout(0xc000139050, 0xc00011fd50, 0xc0000b6b10, 0x29)
        /usr/src/go/pkg/mod/github.com/go-git/go-git/v5@v5.2.0/worktree.go:183 +0x244
go-gitgud/gitget.GitGet.Start(0x7ffe874e674d, 0x16, 0x5, 0xc0000c8980, 0x20, 0xc000138000, 0x7ffe874e6754, 0x9, 0x0, 0x0)
        /usr/src/git/go-gitgud/gitget/gitget.go:116 +0xb99
main.main()
        /usr/src/git/go-gitgud/main.go:38 +0x557
exit status 2
```

**TLDR**
Doing a Checkout with force on a repository with both staged commits for adding a file, unstaged deletion of the file, and the files not present will cause a nil pointer dereference. 

**Additional question**
Performing a standard `git checkout` will not result in the files being created, so is it possible to recreate the same `git checkout <file>` functionality in go-git? 

I cannot find an easy way to re-create the repository scenario, however if someone more familiar with git internals can't suggest an easy way of doing it, I can modify an existing bash based tool to produce it if needs be.
